src/Microsoft.ML.Data/Transforms/KeyToVectorTransform.cs

```diff
@@ -206,393 +186,609 @@ public override void Save(ModelSaveContext ctx)
-            Contracts.AssertValue(info);
-            Contracts.Assert(info.TypeSrc.ItemType.IsKey);
-            Contracts.AssertValue(md);
-
-            int size = info.TypeSrc.ItemType.KeyCount;
-            Contracts.Assert(size > 0);
-
-            // See if the source has key names.
-            var typeNames = input.GetMetadataTypeOrNull(MetadataUtils.Kinds.KeyValues, info.Source);
-            if (typeNames == null || !typeNames.IsKnownSizeVector || !typeNames.ItemType.IsText ||
-                typeNames.VectorSize != size)
+            private readonly KeyToVectorTransform _parent;
+            private readonly ColInfo[] _infos;
+            private readonly ColumnType[] _types;
+
+            public Mapper(KeyToVectorTransform parent, ISchema inputSchema)
+                : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
+            {
+                _parent = parent;
+                _infos = _parent.CreateInfos(inputSchema);
+                _types = new ColumnType[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    ColumnType type;
+                    if (_parent._valueCounts[i] == 1)
+                        type = new VectorType(NumberType.Float, _parent._sizes[i]);
+                    else
+                        type = new VectorType(NumberType.Float, _parent._valueCounts[i], _parent._sizes[i]);
+                    _types[i] = type;
+                }
+            }
+
+            public override RowMapperColumnInfo[] GetOutputColumns()
             {
-                typeNames = null;
+                var result = new RowMapperColumnInfo[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out int colIndex);
+                    Host.Assert(colIndex >= 0);
+                    var colMetaInfo = new ColumnMetadataInfo(_parent.ColumnPairs[i].output);
+                    AddMetadata(i, colMetaInfo);
+
+                    ColumnType type;
+                    if (_parent._valueCounts[i] == 1)
+                        type = new VectorType(NumberType.Float, _parent._sizes[i]);
+                    else
+                        type = new VectorType(NumberType.Float, _parent._valueCounts[i], _parent._sizes[i]);
+                    result[i] = new RowMapperColumnInfo(_parent.ColumnPairs[i].output, _types[i], colMetaInfo);
+                }
+                return result;
             }
 
-            // Don't pass through any source column metadata.
-            using (var bldr = md.BuildMetadata(iinfo))
+            private void AddMetadata(int i, ColumnMetadataInfo colMetaInfo)
             {
-                if (bag || info.TypeSrc.ValueCount == 1)
+                InputSchema.TryGetColumnIndex(_infos[i].Source, out int srcCol);
+                //IVAN: Simplify
```

Comment:
>IVAN [](start = 18, length = 4)

get rid of //IVAN

Reply:
Don't get rid of Ivan. He's such a useful, competent person to have around.

---
In reply to: [216100588](https://github.com/dotnet/machinelearning/pull/858#discussion_r216100588) [](ancestors = 216100588)